02_maps/04/main.go

```diff
@@ -0,0 +1,132 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+)
+
+type Student struct {
+	Id    int    `json:"id"`
+	Name  string `json:"name"`
+	Grade int    `json:"grade"`
+}
+
+type Object struct {
+	Id   int    `json:"id"`
+	Name string `json:"name"`
+}
+
+type Result struct {
+	ObjectId  int `json:"object_id"`
+	StudentId int `json:"student_id"`
+	Result    int `json:"result"`
+}
+
+type Data struct {
+	Students []Student `json:"students"`
+	Objects  []Object  `json:"objects"`
+	Results  []Result  `json:"results"`
+}
+
+func main() {
+	printAggregatedResultTable()
+}
+
+/*
+*
+4. Для предыдущей задачи необходимо вывести сводную таблицу по всем предметам
+*/
+
+type subjectGrade struct {
+	grade  int
+	result []int
+}
+
+func printAggregatedResultTable() {
+	data := readFile()
+
+	subjectsResult := make(map[string]map[int]subjectGrade)
+
+	for _, result := range data.Results {
+		student := getStudentById(result.StudentId, data.Students)
+		objectName := getObjectNameById(result.ObjectId, data.Objects)
+
+		value, ok := subjectsResult[objectName]
+		if !ok {
+			value = make(map[int]subjectGrade)
+			subjectsResult[objectName] = value
+		}
+
+		subjectGradeValue, ok := value[student.Grade]
+		if ok {
+			subjectGradeValue.result = append(subjectGradeValue.result, result.Result)
+		} else {
+			subjectGradeValue = subjectGrade{
+				grade:  student.Grade,
+				result: []int{result.Result},
+			}
+		}
```

Comment:
даже если у тебя еще нет мапе subjectGradeValue, ты всегда можешь записать туда как будто есть, потому что на переменную subjectGradeValue уже выделена память, и спокойно можно к ней обращаться, а append проинициализирует слайс за тебя 
```suggestion
		subjectGradeValue.result = append(subjectGradeValue.result, result.Result)
		subjectGradeValue.grade =  student.Grade
```